polynomial.hpp

```diff
@@ -0,0 +1,220 @@
+        p += q;
+        return p;
+    }
+    template <typename T>
+    inline auto operator-(polynomial<T> p, polynomial<T> q) {
+    	p -= q;
+    	return p;
+    }
+    template <typename T>
+    inline auto operator*(polynomial<T> p, polynomial<T> q) {
+    	p *= q;
+    	return p;
+    }
+    template <typename T>
+    inline auto operator/(polynomial<T> p, polynomial<T> q) {
+    	p /= q;
+    	return p;
+    }
+    template <typename T>
+    inline auto operator%(polynomial<T> p, polynomial<T> q) {
+    	p %= q;
+    	return p;
+    }
+    template <typename T>
+    inline auto divide(polynomial<T> p, polynomial<T> q)
+    ->std::pair<polynomial<T>, polynomial<T>> {
+    	polynomial<T> quotient, remainder;
+        quotient = p / q;
+        remainder = p % q;
+        return std::make_pair(quotient, remainder);
+    }
+    template <typename T>
+    inline auto root_rational_candidates(polynomial<T> p) -> std::vector<T>{
+        size_t P = p.coefficients[p.degree];
+        size_t Q = p.coefficients[0];
+        std::vector<T> factors_of_P{};
+        std::vector<T> factors_of_Q{};
+        std::vector<T> factors{};
+        for(std::size_t i = 1; i < P; i++){
+            if(P % i == 0){
+                factors_of_P.push_back(i);
+            }
+        }
+        for(std::size_t i = 1; i < Q; i++){
+            if(Q % i == 0){
+                factors_of_Q.push_back(i);
+            }
+        }
+        size_t duplicates = 0;
+        for(std::size_t i = 0; i < factors_of_P.size; i++){
+            for(std::size_t j = 0; j < factors_of_Q.size; j++){
+                for(std::size_t k = 0; k < factors.size; k++){
+                    if(factors_of_P[i] / factors_of_Q[j] == factors[k]){
+                        duplicates = 1;
+                        break;
+                    }
+                }
+                if(duplicates == 0){
+                    factors.push_back(factors_of_P[i] / factors_of_Q[j]);
+                }
+                duplicates = 0;
+            }
+        }
+        size_t results_positive = 0;
+        size_t results_negative = 0;
+        for(std::size_t i = 0; i < factors.size; i++){
+            for(std::size_t j = 0; j < p.degree; j++){
+                results_positive += p.coefficients[j] * factors[i];
+                results_negative += p.coefficients[j] * -factors[i];
+            }
+            if(results_positive == 0){
+                return factors[i];
+            }else if(results_negative == 0){
+                return -factors[i];
+            }
+            results_positive = 0;
+            results_negative = 0;
+        }
+        throw std::invalid_argument("Polynomial has no rational roots");
+    }
+    template <typename T>
+    inline auto gcd(polynomial<T> p, polynomial<T> q) -> std::pair<polynomial<T>, std::vector<T>>
+    {
+	std::pair<polynomial<T>, std::vector<T>> result;
+	polynomial<T> x_prev{1}, y_prev{0}, x{0}, y{1};
+	std::vector<T> results;
```

Comment:
no need for vector of results... you just have to remember last two terms in the recursion equations